banyand/stream/service_test.go

```diff
@@ -0,0 +1,302 @@
+}
+
+type entityEventMatcher struct {
+	action databasev1.Action
+}
+
+func (s *entityEventMatcher) Matches(x interface{}) bool {
+	if m, messageOk := x.(bus.Message); messageOk {
+		if evt, dataOk := m.Data().(*databasev1.EntityEvent); dataOk {
+			return evt.Action == s.action
+		}
+	}
+
+	return false
+}
+
+func (s *entityEventMatcher) String() string {
+	return fmt.Sprintf("entity-event-matcher(%s)", databasev1.Action_name[int32(s.action)])
+}
+
+type streamEventSubscriber struct {
+	repo                 discovery.ServiceRepo
+	shardEventSubscriber bus.MessageListener
+	entityEventListener  bus.MessageListener
+}
+
+func (ses *streamEventSubscriber) Name() string {
+	return "stream-event-subscriber"
+}
+
+func (ses *streamEventSubscriber) PreRun() error {
+	components := []struct {
+		shardEvent  bus.Topic
+		entityEvent bus.Topic
+	}{
+		{
+			shardEvent:  event.StreamTopicShardEvent,
+			entityEvent: event.StreamTopicEntityEvent,
+		},
+	}
+	for _, c := range components {
+		err := ses.repo.Subscribe(c.shardEvent, ses.shardEventSubscriber)
+		if err != nil {
+			return err
+		}
+		err = ses.repo.Subscribe(c.entityEvent, ses.entityEventListener)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// service to preload stream
+type preloadStreamService struct {
+	metaSvc metadata.Service
+}
+
+func (p *preloadStreamService) Name() string {
+	return "preload-stream"
+}
+
+func (p *preloadStreamService) PreRun() error {
+	return teststream.PreloadSchema(p.metaSvc.SchemaRegistry())
+}
+
+var ctrl *gomock.Controller
+
+// BeforeSuite - Init logger
+var _ = BeforeSuite(func() {
+	Expect(logger.Init(logger.Logging{
+		Env:   "dev",
+		Level: "info",
+	})).To(Succeed())
+})
+
+// BeforeEach - Create Mock Controller
+var _ = BeforeEach(func() {
+	ctrl = gomock.NewController(GinkgoT())
+	Expect(ctrl).ShouldNot(BeNil())
+})
+
+var _ = Describe("Stream Service", func() {
+	var closer run.Service
+	var streamService Service
+	var metadataService metadata.Service
+	var shardEventListener *bus.MockMessageListener
+	var entityEventListener *bus.MockMessageListener
+
+	BeforeEach(func() {
+		var flags []string
+
+		// Init Discovery
+		repo, err := discovery.NewServiceRepo(context.TODO())
+		Expect(err).NotTo(HaveOccurred())
+
+		// Init Pipeline
+		pipeline, err := queue.NewQueue(context.TODO(), repo)
+		Expect(err).NotTo(HaveOccurred())
+
+		// Init Metadata Service
+		metadataService, err = metadata.NewService(context.TODO())
+		Expect(err).NotTo(HaveOccurred())
+		etcdRootDir := teststream.RandomTempDir()
+		flags = append(flags, "--metadata-root-path="+etcdRootDir)
+		DeferCleanup(func() {
+			_ = os.RemoveAll(etcdRootDir)
+		})
+
+		// Init Stream Service
+		streamService, err = NewService(context.TODO(), metadataService, repo, pipeline)
+		Expect(err).NotTo(HaveOccurred())
+		rootPath, deferFunc, err := test.NewSpace()
+		Expect(err).NotTo(HaveOccurred())
+		DeferCleanup(func() {
+			deferFunc()
```

Comment:
The testing temp dir is removed before the stream module managed in the run-group tends to close the kv storage.
This will cause a panic thrown on the kv closing.

Reply:
Fixed